src/coprocessor/codec/mysql/duration.rs

```diff
@@ -253,97 +254,222 @@ mod parser {
+    }
+
+    /// Raw transmutation from u64.
+    pub fn from_bits(v: u64) -> Result<Duration> {
+        let mut duration = Duration(v);
+
+        check_micros(duration.micros())?;
+        check_second(duration.secs())?;
+        check_minute(duration.minutes())?;
+        check_hour(duration.hours())?;
+
+        duration.set_reserved(false);
+        Ok(duration)
     }
 
-    pub fn nano_secs(&self) -> u32 {
-        self.dur.subsec_nanos()
+    /// Returns the identity element of `Duration`
+    pub fn zero() -> Duration {
+        Duration(0)
+    }
+
+    /// Returns true if self is equal to the additive identity.
+    pub fn is_zero(mut self) -> bool {
+        self.set_neg(false);
+        self.set_fsp(0);
+        self.to_bits() == 0
+    }
+
+    /// Returns the absolute value of `Duration`
+    pub fn abs(mut self) -> Self {
+        self.set_neg(false);
+        self
+    }
+
+    /// Returns the fractional part of `Duration`, in whole microseconds.
+    pub fn subsec_micros(self) -> u32 {
+        self.micros()
+    }
+
+    /// Returns the number of whole seconds contained by this Duration.
+    pub fn to_secs(self) -> i32 {
+        let secs =
+            (self.hours() * SECS_PER_HOUR + self.minutes() * SECS_PER_MINUTE + self.secs()) as i32;
+
+        if self.get_neg() {
+            -secs
+        } else {
+            secs
+        }
     }
 
-    pub fn to_secs(&self) -> f64 {
-        let res = self.dur.as_secs() as f64 + f64::from(self.dur.subsec_nanos()) * 10e-9;
-        if self.neg {
+    /// Returns the number of seconds contained by this Duration as f64.
+    /// The returned value does include the fractional (nanosecond) part of the duration.
+    pub fn to_secs_f64(self) -> f64 {
+        let res = f64::from(self.to_secs().abs()) + f64::from(self.subsec_micros()) * 1e-6;
+
+        if self.get_neg() {
             -res
         } else {
             res
         }
     }
 
-    pub fn is_zero(&self) -> bool {
-        self.to_nanos() == 0
-    }
+    /// Returns the `Duration` in whole nanoseconds
+    pub fn to_nanos(self) -> i64 {
+        let nanos =
+            i64::from(self.to_secs().abs()) * NANOS_PER_SEC + i64::from(self.micros() * 1000);
 
-    pub fn to_nanos(&self) -> i64 {
-        let nanos = self.dur.as_secs() as i64 * NANOS_PER_SEC + i64::from(self.dur.subsec_nanos());
-        if self.neg {
+        if self.get_neg() {
             -nanos
         } else {
             nanos
         }
     }
 
+    /// Constructs a `Duration` from `nanos` with `fsp`
     pub fn from_nanos(nanos: i64, fsp: i8) -> Result<Duration> {
+        let fsp = check_fsp(fsp)?;
         let neg = nanos < 0;
-        let nanos = nanos.abs();
 
-        let dur = StdDuration::new(
-            (nanos / NANOS_PER_SEC) as u64,
-            (nanos % NANOS_PER_SEC) as u32,
-        );
-        Duration::new(dur, neg, fsp)
+        let nanos = nanos.checked_abs().ok_or(invalid_type!("nanos overflow"))?;
```

Comment:
We can remove this `checked_`, since nanos will be / 1000 anyway:

```rust
let micros = (nanos / 100).abs();
```

Reply:
what if `nanos` is `i64::MIN`

Reply:
Then micros will be i64::MIN / 1000.